File: app/data_structures/street_status_table.py
from typing import Dict, Optional
from app.models.street import Street
from app.models.congestion_level import CongestionLevel

class StreetStatusTable:
    def __init__(self):
        self._street_map: Dict[str, Street] = {}

    def add_street(self, street: Street):
        if street and street.street_name:
            self._street_map[street.street_name] = street

    def get_street_status(self, street_name: str) -> Optional[Street]:
        return self._street_map.get(street_name)

    def update_street_status(self, street_name: str, level: CongestionLevel, has_accident: bool, is_blocked: bool) -> bool:
        street = self.get_street_status(street_name)
        if street:
            street.congestion_level = level
            street.has_accident = has_accident
            street.is_blocked = is_blocked
            return True
        return False # Street not found

    def get_all_streets(self) -> Dict[str, Street]:
        return self._street_map.copy() # Return a copy

File: src/main/java/com/rotainteligente/datastructures/StreetStatusTable.java
package com.rotainteligente.datastructures;

import com.rotainteligente.model.CongestionLevel;
import com.rotainteligente.model.Street;

import java.util.HashMap;
import java.util.Map;

public class StreetStatusTable {
    private Map<String, Street> streetMap;

    public StreetStatusTable() {
        this.streetMap = new HashMap<>();
    }

    public void addStreet(Street street) {
        if (street != null && street.getStreetName() != null) {
            this.streetMap.put(street.getStreetName(), street);
        }
    }

    public Street getStreetStatus(String streetName) {
        return this.streetMap.get(streetName);
    }

    public boolean updateStreetStatus(String streetName, CongestionLevel level, boolean hasAccident, boolean isBlocked) {
        Street street = this.streetMap.get(streetName);
        if (street != null) {
            street.setCongestionLevel(level);
            street.setHasAccident(hasAccident);
            street.setBlocked(isBlocked);
            return true;
        }
        return false; // Street not found
    }

    public Map<String, Street> getAllStreets() {
        return new HashMap<>(this.streetMap); // Return a copy to prevent external modification
    }
}
